rust/worker/src/blockstore/arrow_blockfile/blockfile.rs

```diff
@@ -94,6 +95,48 @@ impl Blockfile for ArrowBlockfile {
         }
     }
 
+    fn delete(&mut self, key: BlockfileKey) -> Result<(), Box<dyn ChromaError>> {
+        if !self.in_transaction() {
+            return Err(Box::new(BlockfileError::TransactionNotInProgress));
+        }
+
+        self.validate_key(&key)?;
+
+        let transaction_state = match &self.transaction_state {
+            None => return Err(Box::new(BlockfileError::TransactionNotInProgress)),
+            Some(transaction_state) => transaction_state,
+        };
+
+        // Note: The code to get the target block as well as get or create the delta is duplicated
+        // in the set and delete methods. This is due to the lock management making it clunky
+        // to abstract this into a separate method. This likely should be refactored
+        // for cleaner reuse, but for now its ok to repeat ourselves.
+
+        // Get the target block id for the key
+        let transaction_sparse_index = transaction_state.sparse_index.lock();
+        let target_block_id = match *transaction_sparse_index {
+            None => self.sparse_index.lock().get_target_block_id(&key),
+            Some(ref index) => index.lock().get_target_block_id(&key),
+        };
+
+        // See if a delta for the target block already exists, if not create a new one and add it to the transaction state
+        // Creating a delta loads the block entirely into memory
+        let delta = match transaction_state.get_delta_for_block(&target_block_id) {
```

Comment:
I am not familiar with the rollback mechanism here. What happens if we need to rollback the delete?

Reply:
it discards the whole in-memory state